float8_experimental/float8_dynamic_linear.py

```diff
@@ -96,8 +107,108 @@ def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
                 "bias": False,
             }
             new_mod = cls(**super_kwargs)
-        new_mod.weight = mod.weight
-        new_mod.bias = mod.bias
         new_mod.forward_config = ScaledMMConfig(emulate, True if not emulate else False)
         new_mod.backward_config = ScaledMMConfig(emulate, False)
+        if config.enable_fsdp_fp8_all_gather:
```

Comment:
how does the UX look like for applying FSDP2?  How do we expect the user to sequence applying float8 and applying FSDP2?

if FSDP2 is applied after float8, should this weight wrapping live somewhere else?

Reply:
looking at one of the test cases,

```
            ref_module = self.swap_linear_with_dynamic(ref_module).cuda()
            with set_enable_fsdp_fp8_all_gather(enable_fsdp_fp8_all_gather):
                module = self.swap_linear_with_dynamic(module)
            for submodule in module:
                fully_shard(submodule)
            fully_shard(module)
```

seems like clean places to put the wrapping would be either (1) in `from_float` but not dependent on FSDP, or (2) inside or after the `fully_shard` function.  

is having it in the config temporary or the long term solution?  Can we move it to (1) after the parameter wrapper works without FSDP?

Reply:
I hope that the current global flag is a temporary solution. The risk for it to _not_ be a temporary solution would be if the performance of fp8 all-gather cannot beat without the fp8 all-gather for all realistic cases. Notably, we probably require the extra call to `precompute_float8_casts` after the optimizer step in order to make the fp8 all-gather performant enough (for dynamic scaling), which means that enabling the fp8 all-gather would require an intrusive change in UX.

For now, I think the flag will help with testing/experimentation. Originally, it was a bool arg to `from_float`, but @bdhirsh brought up the fact that fp8 all-gather makes more sense to be to totally enabled on all float8 linears or totally disabled, suggesting that a global flag may make more sense than an arg to `from_float`.